arbor/backends/gpu/reduce_by_key.hpp

```diff
@@ -1,12 +1,58 @@
 #pragma once
 
 #include <cstdint>
-#include "cuda_atomic.hpp"
-#include "cuda_common.hpp"
+#include "gpu_atomic.hpp"
+#include "gpu_common.hpp"
 
 namespace arb {
 namespace gpu {
 
+// double shuffle for AMD devices
+#ifndef __NVCC__
```

Comment:
All of this can go into two separate headers, with other stuff like atomics.
One header for HIP, another for CUDA, with specializations for nvcc vs. clang-cuda in the cuda file..